### ISSUE TITLE
Feature request: add option to connect to a different OpenOCD port

### ISSUE THREAD

**Issue by MicBiso** [eclipse-embed-cdt/eclipse-plugins#569]
### Description

In the Debugger tab I can choose the GDB, Telnet and Tcl ports, this is fine and it configures the ports when OpenOCD is invoked.
However I cannot choose a different port GDB number but the one that is specified in the option above.
The use case is related to multicore devices with different (and consecutive ports) for the different cores, e.g. core 1 port 3333, core 2 port 3334, and so on.
At the moment I can only connect to the core 1, i.e. port 3333 and if I specify 3334 as GDB port, it will connect to port 3334 but this will still be the 1st core, with 2nd core having port 3334 and so on. 
If I deselect the flag "Start OpenOCD locally" then for sure I can specify the exact port I want GDB to connect to but with the hassle to start OpenOCD manually.

It would be nice to have the possibility to specify the actual GDB port.
I tried to use -ex "target extended-remote localhost:3334" as "other options" but it does not work.

### Versions

* [plug-in version] 6.3.2
* [Eclipse version] 2022-06 (4.24.0)
* [Java version]
* [operating system] Ubuntu 22.04
* [toolchain version] NA

**Comment by TommyMurphyTM1234** [comment]
Your report seems confusing.

This:

> In the Debugger tab I can choose the GDB, Telnet and Tcl ports, this is fine and it configures the ports when OpenOCD is invoked. 

seems to contradict this:

> However I cannot choose a different port GDB number but the one that is specified in the option above. 

> if I specify 3334 as GDB port, it will connect to port 3334 but this will still be the 1st core

Surely this behaviour is outside the scope of the plugin and dictated by something external?
Maybe you need a custom OpenOCD script in order to discriminate between different core targets?

I think you need to explain the issue in more detail because I cannot see that it's obviously an issue with the plugin itself.

**Comment by TommyMurphyTM1234** [comment]
What OpenOCD script are you using to interact with your target?

**Comment by MicBiso** [comment]
What do you mean by script? The configuration file?

![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/52751354/0432c48b-4fe9-4330-beeb-e3800ac1568f)

**Comment by MicBiso** [comment]
![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/52751354/6c9c71ce-50ee-4be7-b368-e05bb7687d46)

In other words, the "Port Number" here should be left configurable, also when the "Start OpenOCD locally" is checked.
In fact if I change the GDB port, the same is reflected there:

![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/52751354/9c6bfea5-03cd-40ef-91c9-12ece9325e0c)

**Comment by jonahgraham** [comment]
I notice you are using Renesas - I think Renesas' [e2studio](https://www.renesas.com/us/en/software-tool/e-studio) contains launch configurations that allow such customizations, including selecting which cores to connect Eclipse to.

**Comment by MicBiso** [comment]
No, that's not e2studio, it's "plain" Eclipse:

![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/52751354/144fd14e-a48c-42ac-8c26-80326cb78522)

**Comment by TommyMurphyTM1234** [comment]
What if you do this:

1. Uncheck _Start OpenOCD locally_
2. Change _Remote Target > Port number_ to 3334
3. Check _Start OpenOCD locally_

Does GDB now use 3334?

**Comment by MicBiso** [comment]
> But, it seems, that when OpenOCD is started locally by the plugin then GDB always connects to GDB OpenOCD Debugging > Debugger > OpenOCD Setup > GDB Port and there is no option to get it to connect to that plus 1 or whatever.

Correct.

> What if you do this:
> 
> Uncheck Start OpenOCD locally
> Change Remote Target > Port number to 3334
> Check Start OpenOCD locally
> Does GDB now use 3334?

Yes.

**Comment by TommyMurphyTM1234** [comment]
![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/14550604/d836029f-5f80-400d-840a-5a8d5f4a6c40)

**Comment by TommyMurphyTM1234** [comment]
> Yes.

Ok - so that's the solution?

**Comment by MicBiso** [comment]
Suboptimal I would say.
If I use the plugin I do not want to start OpenOCD manually myself.

**Comment by TommyMurphyTM1234** [comment]
OK - I'll have to let @ilg-ul or @jonahgraham comment.

**Comment by jonahgraham** [comment]
> In other words, the "Port Number" here should be left configurable, also when the "Start OpenOCD locally" box is checked.

Yes - I think that would solve the problem. By default the port number should track the port in start openocd locally, but there should be a way to override it.

**Comment by MicBiso** [comment]
> If you want a single openocd instance to listen on multiple ports, you configure the first launcher to start it, and the second one to do not start it, but use the second port defined in the Remote Target field.

Isn't this a bit nasty? I just want to connect to the second port without having to first start an instance of the plugin just for this. As said, it would be enough to have the "Port Number" field to be always configurable.

**Comment by MicBiso** [comment]
Perfect summary, thanks!

**Comment by ilg-ul** [comment]
Do you want to start one or two sessions?

If you want to start a single session, but not for the first core, you have to adjust your script to give you access to that core, and make openocd to listen on a single port, which you can configure in the plugin.

I have close to zero experience with openocd, so I have no idea how your script should be changed, but with J-Link, if I remember right, for multi core devices it was possible to start multiple GDB server instances, each with its port, configured for a separate core, and this simple use case was assumed for openocd too.
 
From Tommy's summary, I understand that you want to start openocd locally with a custom multi-port configuration, let it listen on multiple port, and do not disable the Remote Target fields (which prevents out of sync configurations for regular use cases), to allow the GDB client to connect to another port.

This can be done, but is it really a good solution?

**Comment by TommyMurphyTM1234** [comment]
> Do you want to start one or two sessions?

They want to start one - that connects to the "second" core target (Cortex-M33) for which OpenOCD creates a GDB server on port 3334.

> If you want to start a single session, but not for the first core, you have to adjust your script to give you access to that core, and make openocd to listen on a single port, which you can configure in the plugin.

That's one way to do it but invasive to the default target script.

> This can be done, but is it really a good solution?

I don't think that it's a bad solution.

As @jonahgraham said above, by default _Remote Target > Port number_ should match _OpenOCD Setup > GDB port_, but the former is left editable (even if _OpenOCD Setup > Start OpenOCD locally_ is checked) and GDB is launched by the plugin using the former rather than the latter. That allows the flexibility that is needed here but doesn't impact anything else that I can think of.

In short the suggested changes to the plugin are:

1. Always have _Remote Target > Port number_ enabled/editable
2. Always launch GDB to connect to _Remote Target > Port number_ rather than _OpenOCD Setup > GDB Port_

**Comment by MicBiso** [comment]
Thanks @TommyMurphyTM1234! 
I started typing my answer but yours was better than mine!

**Comment by jonahgraham** [comment]
I prototyped up in PR #570 a fix for this that adds this checkbox:

![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/679236/6dfc3cda-1e6a-4c11-b61d-d02f7c17fefe)

Is this a good solution? Or a non-starter?

**Comment by TommyMurphyTM1234** [comment]
> I prototyped up in PR #570 a fix for this that adds this checkbox:
> 
> ![image](https://user-images.githubusercontent.com/679236/244720958-6dfc3cda-1e6a-4c11-b61d-d02f7c17fefe.png)
> 
> Is this a good solution? Or a non-starter?

At the moment I'd be more inclined to do this:

1. _OpenOCD Setup > GDB port_ and _Remote Target > Port number_ default to 3333
3. Changes to _OpenOCD Setup > GDB port_ are reflected/applied to _Remote Target > Port number_
4. Changes to _Remote Target > Port number_ override whatever was previously there (and are not reflected back to _OpenOCD Setup > GDB port_)
5. _Remote Target > Port number_ is always enabled - even if _OpenOCD Setup > Start OpenOCD locally_ is enabled
6. GDB is always launched by the plugin using _Remote Target > Port number_ rather than (the de-facto situation when _OpenOCD Setup > Start OpenOCD locally_ is checked?) _OpenOCD Setup > GDB port_?

**Comment by jonahgraham** [comment]
> If the default for the checkbox is always true, and the value is not stored persistently in the environment to be used as a default for the next launcher, it is ok.

Yes - that is how I did it. Checkbox is always checked on new launches (i.e. doesn't use `fPersistentPreferences`), and is set back to checked if you toggle the start gdbserver to on as well.

> 
> My concern is that creating 'regular' launchers may inadvertently inherit this less used configuration.

Not sure what this means - please expand unless I addressed your issue in the above paragraph.

**Comment by jonahgraham** [comment]
> At the moment I'd be more inclined to do this:

I like that better too. Maybe a tooltip on _Remote Target > Port number_ to explain this too.

**Comment by jonahgraham** [comment]
Before I refactor based on @TommyMurphyTM1234's suggestion I will wait for @ilg-ul's input.

**Comment by TommyMurphyTM1234** [comment]
> > At the moment I'd be more inclined to do this:
> 
> I like that better too. Maybe a tooltip on _Remote Target > Port number_ to explain this too.

My only concern is about if/how any settings get persisted and applied to future launch configurations.
But the way that Eclipse plugins persist settings (or not) across instances has always been a mystery to me.

**Comment by ilg-ul** [comment]
> creating 'regular' launchers may inadvertently inherit this less used configuration

The mechanism of keeping persistent preferences, used by Eclipse in many places, is both helpful and annoying.

The annoying case would be when the user configures the remote target port 3334, runs a debug session, and the next day creates a regular launcher, for a simple device which starts openocd to listen only on 3333, and scratches his head why the plugin timeouts (and asks for support!).

Thus I would favour a solution which explicitly shows in the interface that the Remote Target is out of sync with the locally started server, and be sure this checkbox always defaults to a position that prevents this case.

**Comment by ilg-ul** [comment]
> the way that Eclipse plugins persist settings (or not) across instances has always been a mystery to me

The mystery is not that deep, most values in the graphical interface are saved as persistent preferences and used as defaults when the plugin creates a new instance of the corresponding object. 

In this case, a subsequent OpenOCD launcher will inherit most of the values of the previously created OpenOCD launcher, a subsequent QEMU launcher will inherit most of the values of the previously created QEMU launcher, etc.

**Comment by jonahgraham** [comment]
I'm going to experiment if I can make something without a checkbox as @TommyMurphyTM1234 suggests that also explicitly shows what is going on and defaults as expected.

**Comment by ilg-ul** [comment]
Looks good to me. I'll give it a try on Monday, this weekend I'll not be at my computer.

**Comment by jonahgraham** [comment]
@MicBiso - this change has now been merged and a development build has been created by the CI. Can you test the new feature using the p2 site https://download.eclipse.org/embed-cdt/builds/develop/p2/?

**Comment by MicBiso** [comment]
@jonahgraham - thanks.
I'll give it a go asap.

**Comment by MicBiso** [comment]
It works as expected:

![image](https://github.com/eclipse-embed-cdt/eclipse-plugins/assets/52751354/b7cce3a4-bf18-4543-bcb3-58f32a320015)

Thanks again.

**Comment by ilg-ul** [comment]
Great!

Thank you Jonah for your contribution.

If @TommyMurphyTM1234 has no further suggestions (for this and for #566), I'll prepare a new release in the next few days.
